examples/vision/siamese.py

```diff
@@ -0,0 +1,345 @@
+digits (`1` through `9`), digit `1` - from `0` and `2` through `9`, and so on.
+To carry this out, we will select N random images from class A (for example, for
+digit `0`) and pair it with N random images from another class B (for example,
+for digit `1`). Then, we can repeat this process for all classes of digits
+(until digit `9`). Once we have paired digit `0` with other digits, we can
+repeat this process for the remaining classes for the rest of the digits (from
+`1` until `9`).
+
+"""
+
+def make_pairs(x, y):
+    """
+    Parameters
+    ----------
+    x : list
+        List containing images, each index in this list corresponds to one image
+    y : list
+        List containing labels, each label with datatype of `int`
+
+    Returns
+    -------
+    Tuple containing two numpy arrays as (pair_of_samples, labels), 
+    where pair of samples' shape is (2len(x), 2,n_features_dims) and
+    labels are a binary array of shape (2len(x))
+    """
+
+    num_classes = max(y) + 1
+    digit_indices = [np.where(y == i)[0] for i in range(num_classes)]
+
+    pairs = []
+    labels = []
+
+    for idx1 in range(len(x)):
+        # add a matching example
+        x1 = x[idx1]
+        label1 = y[idx1]
+        idx2 = random.choice(digit_indices[label1])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [1]
+    
+        # add a non-matching example
+        label2 = random.randint(0, num_classes-1)
+        while label2 == label1:
+            label2 = random.randint(0, num_classes-1)
+
+        idx2 = random.choice(digit_indices[label2])
+        x2 = x[idx2]
+        
+        pairs += [[x1, x2]]
+        labels += [0]
+
+    return np.array(pairs), np.array(labels).astype('float32')
+
+pairs_train, labels_train = make_pairs(x_train, y_train)
+pairs_test, labels_test = make_pairs(x_test, y_test)
+
+"""
+## Convert the data into TensorFlow Dataset objects
+<br>
+**pairs_train.shape = (120000, 2, 28, 28)** <br>
+Imagine it as: <br>
+**pairs_train.shape = (120000, pair.shape)** <br>
+<br>
+`pairs_train` contains 120K `pairs` in `axis 0`, shape of each pair
+is (2,28,28) hence `each pair` of `pairs_train` contains one image in its
+`axis 0` (do not confuse it with the `axis 0` of `pairs_train`) and the
+other one in the `axis 1`. We will slice `pairs_train` on its `axix 0`
+followed by desired axis of pair to obtain all images (120K) which belong
+either to the `axis 0` or the `axis 1` of all the pairs of `pairs_train`.
+<br>
+**Note:** Do not confuse axes of `pairs_train` with those of
+`pair within pairs_train`, `pairs_train` have only one axis `axis 0` which
+contain 120K pairs, whereas each `pair within pairs_train` have two axis,
+each for one image of a pair.
+
+"""
+
+
+x_train_1 = pairs_train[:,0]
+x_train_2 = pairs_train[:,1]
+# x_train_1.shape = (120000, 28, 28)
+
+x_test_1, x_test_2 = pairs_test[:,0],pairs_test[:,1]
+
+train_pair = tf.data.Dataset.from_tensor_slices((x_train_1, x_train_2))
```

Comment:
Since your data is already NumPy arrays, you don't need to convert it to datasets. Just keep it as NumPy arrays, and pass them to `fit()`.